src/drivers/fluid_opensles.c

```diff
@@ -0,0 +1,414 @@
+  if (dev->output_mix_object)
+    (*dev->output_mix_object)->Destroy (dev->output_mix_object);
+  if (dev->engine)
+    (*dev->engine)->Destroy (dev->engine);  
+  if (dev->sles_buffer_float)
+    FLUID_FREE(dev->sles_buffer_float);
+  if (dev->sles_buffer_short)
+    FLUID_FREE(dev->sles_buffer_short);
+ 
+  FLUID_FREE(dev);
+}
+
+/*
+ * FIXME:
+ * This causes crash on x86 etc. It should be avoided anyways. 
+ * Only non-callback mode uses this function.
+ */
+void adjust_latency(fluid_opensles_audio_driver_t* dev)
+{
+  struct timespec ts;
+  long current_time, wait_in_theory, time_delta;
+
+  wait_in_theory = 1000000 * dev->period_frames / dev->sample_rate;
+
+  /* compute delta time and update 'next expected enqueue' time */
+  clock_gettime(CLOCK_REALTIME, &ts);
+  current_time = ts.tv_sec * 1000000 + ts.tv_nsec / 1000;
+  time_delta = dev->next_expected_enqueue_time == 0 ? 0 : dev->next_expected_enqueue_time - current_time;
+  if (time_delta == 0)
+  {
+    dev->next_expected_enqueue_time += current_time + wait_in_theory;
+  }
+  else
+  {
+    dev->next_expected_enqueue_time += wait_in_theory;
+  }
+  /* take some sleep only if it's running ahead */
+  if (time_delta > 0)
+  {
+    usleep (time_delta);
+  }
+}
+
+void opensles_callback(SLAndroidSimpleBufferQueueItf caller, void *pContext)
+{
+  fluid_opensles_audio_driver_t* dev = (fluid_opensles_audio_driver_t*) pContext;
+  int err;
+  SLresult result;
+
+  process_fluid_buffer(dev);
+
+  if (dev->is_sample_format_float)
+  {
+    result = (*caller)->Enqueue (
+    dev->player_buffer_queue_interface, dev->sles_buffer_float, dev->period_frames * sizeof (float) * NUM_CHANNELS);
+  }
+  else
+  {
+    result = (*caller)->Enqueue (
+    dev->player_buffer_queue_interface, dev->sles_buffer_short, dev->period_frames * sizeof (short) * NUM_CHANNELS);
+  }
+  if (result != 0)
+  {
+    err = result;
+    /* Do not simply break at just one single insufficient buffer. Go on. */
+  }
+}
+
+static fluid_thread_return_t
+fluid_opensles_audio_run(void* d)
+{
+  fluid_opensles_audio_driver_t* dev = (fluid_opensles_audio_driver_t*) d;
+  short *out_short = dev->sles_buffer_short;
+  float *out_float = dev->sles_buffer_float;
+  int period_frames;
+  int err;
+  SLresult result;
+
+  period_frames = dev->period_frames;
+
+  while (dev->cont)
+  {
+    adjust_latency (dev);
+
+    process_fluid_buffer(dev);
+
+    if (dev->is_sample_format_float)
+    {
+      result = (*dev->player_buffer_queue_interface)->Enqueue (
+      dev->player_buffer_queue_interface, out_float, period_frames * sizeof (float) * NUM_CHANNELS);
+    }
+    else
+    {
+      result = (*dev->player_buffer_queue_interface)->Enqueue (
+      dev->player_buffer_queue_interface, out_short, period_frames * sizeof (short) * NUM_CHANNELS);
+    }
+    if (result != 0)
+    {
+      err = result;
+      /* Do not simply break at just one single insufficient buffer. Go on. */
+    }
+  }	/* while (dev->cont) */
+
+  return NULL;
```

Comment:
return FLUID_THREAD_RETURN_VALUE;

Reply:
Actually I'm removing the entire non-callback code (mentioned in another comment).